src/WCSimPrimaryGeneratorAction.cc

```diff
@@ -1552,6 +1558,63 @@ void WCSimPrimaryGeneratorAction::GeneratePrimaries(G4Event* anEvent)
       SetBeamDir(dir);
       SetBeamPDG(pdg);
 
+    } else if (useHepMC3Evt) {
+
+      G4cout << "Using HepMC3 event" << G4endl;
+      // Check if the WCSimNuHepMC3Reader object has been initiaited yet
+      if (!hepmc3_reader) {
+        hepmc3_reader = new WCSimNuHepMC3Reader(hepmc3_filename, myDetector);
+      }
+
+      // Get the next event from the reader
+      hepmc3_reader->ReadEvent(hepmc3_positionGen);
+
+      // Loop over the particles
+      for (HepMC3::GenParticlePtr part : hepmc3_reader->event.particles()) {
+
+        // Skip nuclear remnants
+        if (part->pid() == 2009900000) {
+          continue;
+        }
+
+        // If particle has status 4 then it is a beam particle. This needs writing out, but not simulating
+        if (part->status() == 4) {
+          // Get direction (momentum) and normalise
+          G4ThreeVector dir(part->momentum().px(), part->momentum().py(), part->momentum().pz());
+          dir.set(dir.x() / dir.mag(), dir.y() / dir.mag(), dir.z() / dir.mag());
+
+          // Set write outs
+          SetBeamPDG(part->pdg_id(), 0);
+          SetBeamEnergy(part->momentum().e());
+          SetBeamDir(dir);
```

Comment:
```suggestion
          SetBeamEnergy(part->momentum().e(), 0);
          SetBeamDir(dir, 0);
```
You explicitly put it in the 0th verex for `SetBeamPDG()` - do the same for the others

(of course if there are multiple vertices, the 0 should be replaced with a vertex counter...)